File: Alkitab/src/main/java/yuku/alkitab/songs/MediaController.java
package yuku.alkitab.songs;

import android.app.Activity;
import androidx.annotation.CallSuper;
import java.lang.ref.WeakReference;
import yuku.alkitab.base.util.AppLog;

public abstract class MediaController {
	static final String TAG = MediaController.class.getSimpleName();

	State state = State.reset;
	String url;
	WeakReference<Activity> activityRef;
	WeakReference<MediaStateListener> mediaStateListenerRef;

	public void setUI(Activity activity, MediaStateListener mediaStateListener) {
		activityRef = new WeakReference<>(activity);
		mediaStateListenerRef = new WeakReference<>(mediaStateListener);
	}

	public void setState(final State newState) {
		AppLog.d(TAG, "@@setState newState=" + newState);
		state = newState;
		updateMediaState();
	}

	public void updateMediaState() {
		if (mediaStateListenerRef == null) return;

		final MediaStateListener mediaStateListener = mediaStateListenerRef.get();
		if (mediaStateListener == null) return;

		mediaStateListener.onControllerStateChanged(state);
	}

	@CallSuper
	public void reset() {
		setState(State.reset);
	}

	public void mediaKnownToExist(String url) {
		setState(State.reset_media_known_to_exist);
		this.url = url;
	}

	public abstract void playOrPause(final boolean playInLoop);

	public abstract int[] getProgress();

	public boolean canHaveNewUrl() {
		return state == State.reset || state == State.reset_media_known_to_exist || state == State.error;
	}

	public enum State {
		reset,
		reset_media_known_to_exist,
		preparing,
		playing,
		paused,
		complete,
		error,
	}
}

File: Alkitab/src/main/java/yuku/alkitab/songs/MediaController.kt
package yuku.alkitab.songs

import android.app.Activity
import androidx.annotation.CallSuper
import yuku.alkitab.base.util.AppLog
import java.lang.ref.WeakReference

private const val TAG = "MediaController"

abstract class MediaController {
    var state = State.reset
        set(value) {
            AppLog.d(TAG, "@@setState newState=$value")
            field = value
            updateMediaState()
        }

    protected var url: String? = null
    protected var activityRef: WeakReference<Activity>? = null

    private var mediaStateListenerRef: WeakReference<MediaStateListener>? = null

    fun setUI(activity: Activity, mediaStateListener: MediaStateListener) {
        activityRef = WeakReference(activity)
        mediaStateListenerRef = WeakReference(mediaStateListener)
    }

    fun updateMediaState() {
        mediaStateListenerRef?.get()?.onControllerStateChanged(state)
    }

    @CallSuper
    open fun reset() {
        state = State.reset
    }

    fun mediaKnownToExist(url: String) {
        state = State.reset_media_known_to_exist
        this.url = url
    }

    abstract fun playOrPause(playInLoop: Boolean)

    abstract fun getProgress(): IntArray

    fun canHaveNewUrl(): Boolean {
        return state == State.reset || state == State.reset_media_known_to_exist || state == State.error
    }

    enum class State {
        reset, reset_media_known_to_exist, preparing, playing, paused, complete, error
    }
}
